internal/store/device.go

```diff
@@ -135,35 +152,79 @@ func (s *DeviceStore) Get(ctx context.Context, orgId uuid.UUID, name string) (*a
 	return &apiDevice, nil
 }
 
-func (s *DeviceStore) CreateOrUpdate(ctx context.Context, orgId uuid.UUID, resource *api.Device) (*api.Device, bool, error) {
+func (s *DeviceStore) CreateOrUpdate(ctx context.Context, orgId uuid.UUID, resource *api.Device, callback DeviceStoreCallback) (*api.Device, bool, error) {
+	log := log.WithReqIDFromCtx(ctx, s.log)
 	if resource == nil {
 		return nil, false, fmt.Errorf("resource is nil")
 	}
 	device := model.NewDeviceFromApiResource(resource)
+	if device.Name == "" {
+		return nil, false, fmt.Errorf("resource has no name")
+	}
 	device.OrgID = orgId
 
-	// don't overwrite status
+	// don't overwrite status, generation, or owner
 	device.Status = nil
+	device.Generation = nil
+	device.Owner = nil
 
 	created := false
-	findDevice := model.Device{Resource: model.Resource{Name: *resource.Metadata.Name}}
-	findDevice.OrgID = orgId
-	result := s.db.First(&findDevice)
-	if result.Error != nil {
-		if result.Error == gorm.ErrRecordNotFound {
+	var existingRecord *model.Device
+
+	err := s.db.Transaction(func(innerTx *gorm.DB) (err error) {
+		existingRecord = &model.Device{Resource: model.Resource{OrgID: device.OrgID, Name: device.Name}}
+		result := innerTx.First(existingRecord)
+		log.Printf("db.Find(%s): %d rows affected, error is %v", *existingRecord, result.RowsAffected, result.Error)
+		// NotFound is OK because in that case we will create the record, anything else is a real error
+		if result.Error != nil && !errors.Is(result.Error, gorm.ErrRecordNotFound) {
+			return result.Error
+		}
+		if result.Error != nil {
 			created = true
+			device.Generation = util.Int64ToPtr(1)
+
+			result = innerTx.Create(device)
+			if result.Error != nil {
+				return result.Error
+			}
```

Comment:
can we document why we overwrite result.Error instead of returning it immediately?

Reply:
I made the code clearer to show that the first check of result.Error actually means that the device wasn't found and it's not a real error.